File: javascript/jsfind/tests/findoptions.test.js
/*
 * findoptions_test.js
 *
 * Some tests of findoptions.js
 */

const FindOptions = require('../src/findoptions').FindOptions;
const FindSettings = require('../src/findsettings').FindSettings;

describe('testing findoptions', () => {
    it('testNoArgs', () => {
        const findOptions = new FindOptions();
        findOptions.settingsFromArgs([], (err, settings) => {
            if (err) {
                console.log('There was an error calling settingsFromArgs: ' + err);
                expect(false).toEqual(true);
            }
            expect(settings.archivesOnly).toBeFalsy();
            expect(settings.debug).toBeFalsy();
            expect(settings.excludeHidden).toBeTruthy();
            expect(settings.listDirs).toBeFalsy();
            expect(settings.listFiles).toBeTruthy();
            expect(settings.printUsage).toBeFalsy();
            expect(settings.printVersion).toBeFalsy();
            expect(settings.recursive).toBeTruthy();
            expect(settings.findArchives).toBeFalsy();
            expect(settings.paths.length).toEqual(0);
            expect(settings.verbose).toBeFalsy();
        });
    });

    it('testValidArgs', () => {
        const findOptions = new FindOptions();
        const args = ['-x', 'js,java', '.'];
        findOptions.settingsFromArgs(args, (err, settings) => {
            if (err) {
                console.log('There was an error calling settingsFromArgs: ' + err);
                expect(false).toEqual(true);
            }
            expect(settings.inExtensions.length).toEqual(2);
            expect(settings.inExtensions[0]).toEqual('js');
            expect(settings.inExtensions[1]).toEqual('java');
        });
    });

    it('testArchivesOnly', () => {
        const findOptions = new FindOptions();
        const args = ['--archivesonly'];
        findOptions.settingsFromArgs(args, (err, settings) => {
            if (err) {
                console.log('There was an error calling settingsFromArgs: ' + err);
                expect(false).toEqual(true);
            }
            expect(settings.archivesOnly).toEqual(true);
            expect(settings.includeArchives).toEqual(true);
        });
    });

    it('testDebug', () => {
        const findOptions = new FindOptions();
        const args = ['--debug'];
        findOptions.settingsFromArgs(args, (err, settings) => {
            if (err) {
                console.log('There was an error calling settingsFromArgs: ' + err);
                expect(false).toEqual(true);
            }
            expect(settings.debug).toEqual(true);
            expect(settings.verbose).toEqual(true);
        });
    });

    it('testMissingArg', () => {
        const findOptions = new FindOptions();
        const args = ['-x'];
        findOptions.settingsFromArgs(args, (err) => {
            if (err) {
                const expected = 'Missing argument for option x';
                expect(err.message).toEqual(expected);
            } else {
                console.log('Did not get expected missing argument error');
                expect(false).toEqual(true);
            }
        });
    });

    it('testIvalidArg', () => {
        const findOptions = new FindOptions();
        const args = ['-Q'];
        findOptions.settingsFromArgs(args, (err) => {
            if (err) {
                const expected = 'Invalid option: Q';
                expect(err.message).toEqual(expected);
            } else {
                console.log('Did not get expected unknown option error');
                expect(false).toEqual(true);
            }
        });
    });

    it('testSettingsFromJson', () => {
        const findOptions = new FindOptions();
        const settings = new FindSettings();
        const json = '{\n' +
            '  "path": "~/src/xfind/",\n' +
            '  "in-ext": ["js","ts"],\n' +
            '  "out-dirpattern": "node_module",\n' +
            '  "out-filepattern": ["temp"],\n' +
            '  "debug": true,\n' +
            '  "includehidden": true\n' +
            '}';
        const err = findOptions.settingsFromJson(json, settings);
        expect(err).toBeNull();
        expect(settings.paths.length).toEqual(1);
        expect(settings.paths[0]).toEqual('~/src/xfind/');
        expect(settings.inExtensions.length).toEqual(2);
        expect(settings.outDirPatterns.length).toEqual(1);
        expect(settings.outDirPatterns[0].source).toEqual('node_module');
        expect(settings.outFilePatterns.length).toEqual(1);
        expect(settings.outFilePatterns[0].source).toEqual('temp');
        expect(settings.debug).toBeTruthy();
        expect(settings.verbose).toBeTruthy();
        expect(settings.excludeHidden).toBeFalsy();
    });
});

File: go/gofind/pkg/gofind/findoptions_test.go
package gofind

import "testing"

func TestFindSettingsFromNoArgs(t *testing.T) {
	findOptions := NewFindOptions()

	args := []string{}

	settings, err := findOptions.FindSettingsFromArgs(args)
	if err != nil {
		t.Errorf("FindSettingsFromArgs: err: %v", err)
	}

	if settings.ArchivesOnly() ||
		settings.Debug() ||
		settings.IncludeArchives() ||
		settings.IncludeHidden() ||
		settings.PrintDirs() ||
		!settings.PrintFiles() ||
		settings.PrintUsage() ||
		settings.PrintVersion() ||
		!settings.Recursive() ||
		settings.Verbose() {
		t.Errorf("settings did not match defaults")
	}
}

func TestFindSettingsFromValidArgs(t *testing.T) {
	findOptions := NewFindOptions()

	args := []string{
		"-x", "go", ".",
	}

	settings, err := findOptions.FindSettingsFromArgs(args)
	if err != nil {
		t.Errorf("FindSettingsFromArgs: err: %v", err)
	}

	if settings.Paths()[0] != "." {
		t.Errorf("settings.Paths[0] != \".\"")
	}

	if len(settings.InExtensions()) != 1 {
		t.Errorf("len(settings.InExtensions) = %d, expected 1",
			len(settings.InExtensions()))
	}
	expectedExt := "go"
	if settings.InExtensions()[0] != expectedExt {
		t.Errorf("settings.InExtensions[0] (\"%s\") != \"%s\"",
			settings.InExtensions()[0], expectedExt)
	}
}

func TestFindSettingsFromJson(t *testing.T) {
	findOptions := NewFindOptions()

	jsonSettings := []byte(`{
  "path": "~/src/xfind/",
  "in-ext": ["js","ts"],
  "out-dirpattern": "node_module",
  "out-filepattern": ["temp"],
  "debug": true,
  "includehidden": true
}`)

	settings := GetDefaultFindSettings()
	var err error
	err = findOptions.SettingsFromJson(jsonSettings, settings)
	if err != nil {
		t.Errorf("TestFindSettingsFromJson: err: %v", err)
	}

	if len(settings.Paths()) != 1 {
		t.Errorf("len(settings.Paths) = %d, expected 1",
			len(settings.Paths()))
	}
	if settings.Paths()[0] != "~/src/xfind/" {
		t.Errorf("settings.Paths[0] != \"~/src/xfind/\"")
	}

	if len(settings.InExtensions()) != 2 {
		t.Errorf("len(settings.InExtensions) = %d, expected 2",
			len(settings.InExtensions()))
	}
	expectedInExts := []string{"js", "ts"}
	for i, _ := range expectedInExts {
		if settings.InExtensions()[i] != expectedInExts[i] {
			t.Errorf("settings.InExtensions[%d] (\"%s\") != \"%s\"",
				i, settings.InExtensions()[i], expectedInExts[i])
		}
	}

	if len(settings.OutDirPatterns().patterns) != 1 {
		t.Errorf("len(settings.OutDirPatterns.patterns) = %d, expected 1",
			len(settings.OutDirPatterns().patterns))
	}
	if settings.OutDirPatterns().patterns[0].String() != "node_module" {
		t.Errorf("settings.OutDirPatterns.patterns[0].String() (\"%s\") != \"node_module\"",
			settings.OutDirPatterns().patterns[0].String())
	}

	if len(settings.OutFilePatterns().patterns) != 1 {
		t.Errorf("len(settings.OutFilePatterns.patterns) = %d, expected 1",
			len(settings.OutFilePatterns().patterns))
	}
	if settings.OutFilePatterns().patterns[0].String() != "temp" {
		t.Errorf("settings.OutFilePatterns.patterns[0].String() (\"%s\") != \"temp\"",
			settings.OutFilePatterns().patterns[0].String())
	}

	if !settings.Debug() {
		t.Errorf("settings.Debug (%t) != true", settings.Debug())
	}

	if !settings.IncludeHidden() {
		t.Errorf("settings.IncludeHidden (%t) != true", settings.IncludeHidden())
	}
}
